src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/DefaultHybridCache.L2.cs

```diff
@@ -0,0 +1,153 @@
+
+        static async Task<BufferChunk> AwaitedLegacyAsync(Task<byte[]?> pending, DefaultHybridCache @this)
+        {
+            var bytes = await pending.ConfigureAwait(false);
+            return @this.GetValidPayloadSegment(bytes);
+        }
+
+        static async Task<BufferChunk> AwaitedBuffersAsync(ValueTask<bool> pending, RecyclableArrayBufferWriter<byte> writer)
+        {
+            BufferChunk result = await pending.ConfigureAwait(false)
+                    ? new(writer.DetachCommitted(out var length), length, returnToPool: true)
+                    : default;
+            writer.Dispose(); // it is not accidental that this isn't "using"; avoid recycling if not 100% sure what happened
+            return result;
+        }
+    }
+
+    internal ValueTask SetL2Async(string key, in BufferChunk buffer, HybridCacheEntryOptions? options, CancellationToken token)
+    {
+        Debug.Assert(buffer.Array is not null, "array should be non-null");
+        switch (GetFeatures(CacheFeatures.BackendCache | CacheFeatures.BackendBuffers))
+        {
+            case CacheFeatures.BackendCache: // legacy byte[]-based
+                var arr = buffer.Array!;
+                if (arr.Length != buffer.Length)
+                {
+                    // we'll need a right-sized snapshot
+                    arr = buffer.ToArray();
+                }
+
+                return new(_backendCache!.SetAsync(key, arr, GetOptions(options), token));
+            case CacheFeatures.BackendCache | CacheFeatures.BackendBuffers: // ReadOnlySequence<byte>-based
+                var cache = Unsafe.As<IBufferDistributedCache>(_backendCache!); // type-checked already
+                return cache.SetAsync(key, buffer.AsSequence(), GetOptions(options), token);
+        }
+
+        return default;
+    }
+
+    internal void SetL1<T>(string key, CacheItem<T> value, HybridCacheEntryOptions? options)
+    {
+        // incr ref-count for the the cache itself; this *may* be released via the NeedsEvictionCallback path
+        if (value.TryReserve())
+        {
+            // based on CacheExtensions.Set<TItem>, but with post-eviction recycling
+            using var cacheEntry = _localCache.CreateEntry(key);
+            cacheEntry.AbsoluteExpirationRelativeToNow = options?.LocalCacheExpiration ?? _defaultLocalCacheExpiration;
+            cacheEntry.Value = value;
+            if (value.NeedsEvictionCallback)
+            {
+                _ = cacheEntry.RegisterPostEvictionCallback(CacheItem.SharedOnEviction);
+            }
+        }
+    }
+
+    private BufferChunk GetValidPayloadSegment(byte[]? payload)
+    {
+        if (payload is not null)
+        {
+            if (payload.Length > MaximumPayloadBytes)
+            {
+                ThrowPayloadLengthExceeded(payload.Length);
+            }
+
+            return new(payload);
+        }
+
+        return default;
+    }
+
+    [DoesNotReturn]
+    [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Why no inlining? All throw helpers are marked with aggressive inlining (see https://github.com/dotnet/extensions/blob/main/src/Shared/Throw/Throw.cs).

Reply:
Ah! Thank you for explaining.